crates/c-api/src/wasi.rs

```diff
@@ -183,13 +185,16 @@ pub unsafe extern "C" fn wasi_config_preopen_dir(
         None => return false,
     };
 
-    config
-        .builder
-        .preopened_dir(
-            host_path,
-            guest_path,
-            wasmtime_wasi::DirPerms::all(),
-            wasmtime_wasi::FilePerms::all(),
-        )
+    let dir_perms = wasmtime_wasi::DirPerms::from_bits(dir_perms);
+    let file_perms = wasmtime_wasi::FilePerms::from_bits(file_perms);
+
+    dir_perms
+        .and_then(|dir_perms| {
+            file_perms.map(|file_perms| {
+                config
+                    .builder
+                    .preopened_dir(host_path, guest_path, dir_perms, file_perms)
+            })
+        })
```

Comment:
This feels clunky, but I'm not a rust expert so if you know of a more idiomatic way of handling the error cases, I'd be happy to clean this up.

Reply:
Oh for this I might recommend:

```rust
let dir_perms = wasmtime_wasi::DirPerms::from_bits(dir_perms)
    .ok_or_else(|| anyhow!("invalid directory permissions"))?;
```

That way you can avoid the nested closures

Reply:
That's certainly cleaner! But `wasi_config_preopen_dir` returns `bool` rather than a `Result`, so it doesn't appear that you can use the `?` operator. Is there a way to do an unwrap / early return like that in a function that returns `bool`?
 Or would it be better to change the function signature? Returning strings from a C API would require callers to `free` the returned result, but we could potentially use a numerical error code instead.

Reply:
Ah excellent point! I forgot this returns `bool`, so sorry about that.

For that I'd recommend a `match`-per-translation-of-config to return `false` early and that way it doesn't have to nest.

I do like the idea of changing the function signature too. It's overdue for the WASI APIs here. In doing so though it might be best to rename and shuffle things around. For example the original intention of `wasi.h` was to become a standard API like the wasm-c-api proposal (which hasn't finished being standardized yet). In lieu of that nowadays it's probably best to rename things to `wasmtime_wasi_*`, move the header to `wasmtime/wasi.h`, and update idioms to match the rest of the C API (e.g. `wasmtime_error_t*` instead of `bool`). That's a bit of a larger project though so how about we keep the `bool` for now and defer this refactoring for later?

Reply:
Thanks, that's much cleaner. That's also the pattern already being used for the other parameters of the function, which I somehow didn't recognize 🤦. Sorry for the churn.